ttnn/cpp/ttnn/simple_device.hpp

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "tt_metal/distributed/mesh_device.hpp"
+#include "tt_metal/impl/device/device.hpp"
+
+namespace ttnn {
+
+// SimpleDevice is a wrapper around Device / MeshDevice to use in interfaces that can accept either.
+// This class is cheaply copyable, use value semantics to pass it around.
+//
+// TODO: the eventual goal is to lower this primitive into tt_metal. In the long term, we also want to extend the
+// functionality with the "distributed device" semantics.
+class SimpleDevice {
+   public:
+    // Allow implicit conversion for transparent migration.
+    // Expect the pointers to be non-null, and remain valid for the lifetime of SimpleDevice.
+    SimpleDevice(tt::tt_metal::Device* device) : metal_device_{device} {}
+    SimpleDevice(tt::tt_metal::distributed::MeshDevice* mesh_device) : metal_device_{mesh_device} {}
+    SimpleDevice(const SimpleDevice&) = default;
+    SimpleDevice& operator=(const SimpleDevice&) = default;
+
+    std::vector<tt::tt_metal::Device*> get_devices() {
+        if (auto* device = std::get_if<tt::tt_metal::Device*>(&metal_device_); device != nullptr) {
+            return {*device};
+        } else {
+            return std::get<tt::tt_metal::distributed::MeshDevice*>(metal_device_)->get_devices();
+        }
+    }
+
+   private:
+    std::variant<tt::tt_metal::Device*, tt::tt_metal::distributed::MeshDevice*> metal_device_;
+};
+
+class OptionalSimpleDevice {
```

Comment:
what is the main reason to create another entity? can it be substituted with using? 

`using OptionalSimpleDevice = std::optional<...>;`

Reply:
Added this as a stop gap to be able to bind `std::optional<std::reference_wrapper<Device>>` and other flavors implicitly (for pybind). The long term solution probably involves moving ttnn onto the single "device" API, so that we can remove some code from this then.

I did move `OptionalSimpleDevice` into the creation header though, the only place where I think we need the reference wrapper bindings.